sensugrid.py

```diff
@@ -95,21 +101,48 @@ def _cmp(x, y):
 
 
 def get_agg_data(dc):
-    r = agg_data(dc, get_data(dc, timeout), get_stashes(dc, timeout))
-    return r
+    try:
+        r = agg_data(dc, cache["{0}_data".format(dc)], cache["{0}_stashes".format(dc)])
+        return r, cache["{0}_filters".format(dc)]
+    except KeyError:
+        data = list()
+        stashes = list()
+        data_futures = list()
+        stashes_futures = list()
+        filters_futures = list()
+        data_executor = concurrent.futures.ThreadPoolExecutor(1)
+        data_futures.append(data_executor.submit(get_data, dc, timeout))
+        stashes_executor = concurrent.futures.ThreadPoolExecutor(1)
+        stashes_futures.append(stashes_executor.submit(get_stashes, dc, timeout))
+        filters_executor = concurrent.futures.ThreadPoolExecutor(1)
+        filters_futures.append(filters_executor.submit(filter_data, timeout, dc))
+        for future_result in as_completed(data_futures):
+            data = future_result.result()
+        for future_result in as_completed(stashes_futures):
+            stashes = future_result.result()
+        for future_result in as_completed(filters_futures):
+            filters = future_result.result()
+
+        cache.set("{0}_data".format(dc), data, expire=cache_expire_time)
+        cache.set("{0}_stashes".format(dc), stashes, expire=cache_expire_time)
+        cache.set("{0}_filters".format(dc), filters, expire=cache_expire_time)
+
+        r = agg_data(dc, data, stashes)
+        return r, filters
 
 
 @app.route('/', methods=['GET'])
 def root():
     aggregated = list()
-    pool = ThreadPool(len(dcs))
-    try:
-        aggregated = pool.map(get_agg_data, dcs)
-    except Exception as e:
-        print("Exception: ", e)
-    finally:
-        pool.close()
-    return render_template('data.html', dcs=dcs, data=aggregated, filter_data=get_filter_data(dcs, timeout), appcfg=appcfg)
+    futures = list()
+    filters = list()
+    agg_data_executor = concurrent.futures.ThreadPoolExecutor(len(dcs))
+    [futures.append(agg_data_executor.submit(get_agg_data, dc)) for dc in dcs]
```

Comment:
i agree that list comprehension seems more elegant, but it's less readable; can this be the normal
```
for dc in dcs:
    futures.append......
```
?